app/models/internal/impl/DefaultRenderedReport.java

```diff
@@ -0,0 +1,165 @@
+ * limitations under the License.
+ */
+
+package models.internal.impl;
+
+import com.arpnetworking.commons.builder.OvalBuilder;
+import com.arpnetworking.metrics.portal.reports.RenderedReport;
+import com.google.common.base.MoreObjects;
+import models.internal.reports.ReportFormat;
+import net.sf.oval.constraint.NotNull;
+
+import java.io.ByteArrayInputStream;
+import java.io.InputStream;
+import java.time.Instant;
+import java.util.Arrays;
+import java.util.Objects;
+
+/**
+ * A document containing the results of a particular rendering of a particular {@link models.internal.reports.Report}.
+ *
+ * @author Spencer Pearson (spencerpearson at dropbox dot com)
+ */
+public final class DefaultRenderedReport implements RenderedReport {
+    @Override
+    public InputStream getBytes() {
+        return new ByteArrayInputStream(_bytes);
+    }
+
+    @Override
+    public Instant getScheduledFor() {
+        return _scheduledFor;
+    }
+
+    @Override
+    public Instant getGeneratedAt() {
+        return _generatedAt;
+    }
+
+    @Override
+    public ReportFormat getFormat() {
+        return _format;
+    }
+
+    @Override
+    public boolean equals(final Object o) {
+        if (this == o) {
+            return true;
+        }
+        if (!(o instanceof DefaultRenderedReport)) {
+            return false;
+        }
+        final DefaultRenderedReport that = (DefaultRenderedReport) o;
+        return Arrays.equals(_bytes, that._bytes)
+                && _scheduledFor.equals(that._scheduledFor)
+                && _generatedAt.equals(that._generatedAt)
+                && _format.equals(that._format);
+    }
+
+    @Override
+    public int hashCode() {
+        int result = Objects.hash(_scheduledFor, _generatedAt, _format);
+        result = 31 * result + Arrays.hashCode(_bytes);
+        return result;
+    }
+
+    @Override
+    public String toString() {
+        return MoreObjects.toStringHelper(this)
+                .add("_bytes", _bytes)
+                .add("_scheduledFor", _scheduledFor)
+                .add("_generatedAt", _generatedAt)
+                .add("_format", _format)
+                .toString();
+    }
+
+    private final byte[] _bytes;
+    private final Instant _scheduledFor;
+    private final Instant _generatedAt;
+    private final ReportFormat _format;
+
+    /**
+     * Builder implementation that constructs {@code DefaultReport}.
+     */
+    public static final class Builder extends OvalBuilder<DefaultRenderedReport> {
+        /**
+         * Public Constructor.
+         */
+        public Builder() {
+            super(DefaultRenderedReport::new);
+        }
+
+        /**
+         * Set the report bytes. Required. Cannot be null.
+         *
+         * @param bytes The report bytes.
+         * @return This instance of {@code Builder}.
+         */
+        public Builder setBytes(final byte[] bytes) {
+            _bytes = bytes.clone();
```

Comment:
Nice! Typically we do this in the ctor.

Reply:
Hmm. In this case, due to the mutability issue you point out above, I feel that doing it here is more appropriate. (If I expected that we'd call `setBytes()` on the same instance a _whole lot_, I could be convinced that the performance cost would justify moving the `clone()` into the constructor; but since I don't expect that, I think I'd rather keep as few externally-mutable objects around as possible.)